### PR TITLE
Verifier integration

### PR DESCRIPTION
**Pull request by Alan-Jowett**
This pull request integrates the eBPF verifier as a submodule and updates the `libfuzzer` harness to utilize the verifier. The changes include modifications to submodule configurations, build scripts, and the addition of new verification and debugging functionalities in the `libfuzzer` harness.

### Integration of eBPF Verifier:

* [`.gitmodules`](diffhunk://#diff-fe7afb5c9c916e521401d3fcfb4277d5071798c3baf83baf11d6071742823584R7-R9): Added the `ebpf-verifier` submodule.
* [`CMakeLists.txt`](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR46): Included the `ebpf-verifier` directory in the build when `UBPF_ENABLE_LIBFUZZER` is enabled.
* [`external/ebpf-verifier`](diffhunk://#diff-0f47f98c433f156a9980e12abb7595d356ac1690a2e6ddbc3d51f7a4f0c7621eR1): Added the submodule commit reference for `ebpf-verifier`.

### Build Script Updates:

* [`libfuzzer/CMakeLists.txt`](diffhunk://#diff-1741b861004a93abe82df187da3c41bae7ed269806a0fa9a2ef98eff4e41e04aR16-R34): Updated include directories and linked the `ebpf-verifier` library.

### Enhancements to `libfuzzer` Harness:

* `libfuzzer/libfuzz_harness.cc`: 
  - Included headers for the eBPF verifier and platform.
  - Added `verify_bpf_byte_code` function to verify BPF bytecode using the eBPF verifier.
  - Implemented context management and debug functions for the uBPF context.
  - Integrated verification and debug functions into the interpreter and JIT execution paths. [[1]](diffhunk://#diff-90d0c71847f23efa012693fe27bf93e459ad6a1b101cb48db77dd6205fbaf524R308-R318) [[2]](diffhunk://#diff-90d0c71847f23efa012693fe27bf93e459ad6a1b101cb48db77dd6205fbaf524R341-R342) [[3]](diffhunk://#diff-90d0c71847f23efa012693fe27bf93e459ad6a1b101cb48db77dd6205fbaf524L157-R359)
  - Added a call to `verify_bpf_byte_code` in the main fuzzer test function.

### API Changes:

* [`vm/inc/ubpf.h`](diffhunk://#diff-47050594ea372d3c6a44bce95ada2feb7306a466565cca08aa9458d69cf92b3aL580-R581): Added `register_mask` parameter to the debug function signature.
* [`vm/ubpf_vm.c`](diffhunk://#diff-45bc22963d9274fb910b15adf80be8ac6b9cb8a40b6a57bee0a92f821116b32fL682-R682): Updated the call to the debug function to include `register_mask`.

### CODE REVIEW

File: libfuzzer/libfuzz_harness.cc [hunk 1]
```diff
@@ -10,27 +10,215 @@
...
 #include "ebpf.h"
 #include "ubpf.h"
+#undef ebpf_inst
 }
 
 #include "test_helpers.h"
 #include <cassert>
 
+/**
+ * @brief Context structure passed to the BPF program. Modelled after the context structure used by XDP.
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:29]:
```suggestion
 * @brief Context structure passed to the BPF program. Modeled after the context structure used by XDP.
```
Modelled is british.  Modeled is american.

File: libfuzzer/libfuzz_harness.cc [hunk 2]
```diff
@@ -10,27 +10,215 @@
...
+ebpf_context_descriptor_t g_ebpf_context_descriptor_ubpf = {
+    .size = sizeof(ubpf_context_t),
+    .data = offsetof(ubpf_context_t, data),
+    .end = offsetof(ubpf_context_t, data_end),
+    .meta = -1,
+};
+
+/**
+ * @brief Description of the program type. This is used by the verifier to determine the what the context structure to use as well as the helper functions that are available.
+ *
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:52]:
```suggestion
 * @brief Description of the program type. This is used by the verifier to determine what context structure to use as well as the helper functions that are available.
```

File: libfuzzer/libfuzz_harness.cc [hunk 3]
```diff
@@ -10,27 +10,215 @@
...
+ * @param[in] map_descriptors The map descriptors to populate.
+ * @param[in] data The data in the maps section.
+ * @param[in] map_record_size The size of each map record.
+ * @param[in] map_count The number of maps in the maps section.
+ * @param[in] platform The platform specific data.
+ * @param[in] options Options for the verifier.
+ */
+void ubpf_parse_maps_section(std::vector<EbpfMapDescriptor>& map_descriptors, const char* data,
+                                           size_t map_record_size, int map_count,
+                                           const struct ebpf_platform_t* platform, ebpf_verifier_options_t options)
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:128]:
```suggestion
                                         size_t map_record_size, int map_count,
                                         const struct ebpf_platform_t* platform, ebpf_verifier_options_t options)
```

File: libfuzzer/libfuzz_harness.cc [hunk 4]
```diff
@@ -10,27 +10,215 @@
...
+ */
+EbpfMapDescriptor& ubpf_get_map_descriptor(int map_fd)
+{
+    // Once the fuzzer supports maps, this function should be implemented to return the map descriptor for the given map file descriptor.
+    UNREFERENCED_PARAMETER(map_fd);
+    throw std::runtime_error("get_map_descriptor not implemented");
+}
+
+/**
+ * @brief The platform abstraction for verifier to call into the uBPF fuzzer platform.
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:166]:
```suggestion
 * @brief The platform abstraction for the verifier to call into the uBPF fuzzer platform.
```

File: libfuzzer/libfuzz_harness.cc [hunk 5]
```diff
@@ -10,27 +10,215 @@
...
+/**
+ * @brief Dispatcher for the helper functions.
+ *
+ * @param p0 First parameter to the helper function.
+ * @param p1 Second parameter to the helper function.
+ * @param p2 Third parameter to the helper function.
+ * @param p3 Fourth parameter to the helper function.
+ * @param p4 Fifth parameter to the helper function.
+ * @param idx Index of the helper function to call.
+ * @param cookie Cookie to pass to the helper function.
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:190]:
```suggestion
 * @param[in] p0 First parameter to the helper function.
 * @param[in] p1 Second parameter to the helper function.
 * @param[in] p2 Third parameter to the helper function.
 * @param[in] p3 Fourth parameter to the helper function.
 * @param[in] p4 Fifth parameter to the helper function.
 * @param[in] idx Index of the helper function to call.
 * @param[in] cookie Cookie to pass to the helper function.
```

File: libfuzzer/libfuzz_harness.cc [hunk 6]
```diff
@@ -42,6 +230,64 @@ int null_printf(FILE* stream, const char* format, ...)
...
+        .type = g_ubpf_program_type,
+    };
+    std::string section;
+    std::string file;
+    raw_program raw_prog{file, section, 0, {}, instructions, info};
+
+    // Unpack the program into a sequence of instructions that the verifier can understand.
+    std::variant<InstructionSeq, std::string> prog_or_error = unmarshal(raw_prog);
+    if (!std::holds_alternative<InstructionSeq>(prog_or_error)) {
+        //std::cout << "Failed to unmarshal program : " << std::get<std::string>(prog_or_error) << std::endl;
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:259]:
```suggestion
```

File: libfuzzer/libfuzz_harness.cc [hunk 7]
```diff
@@ -95,6 +341,146 @@ ubpf_vm_ptr create_ubpf_vm(const std::vector<uint8_t>& program_code)
...
+    } else {
+        return address_type_t::Unknown;
+    }
+}
+
+/**
+ * @brief Function invoked prior to executing each instruction in the program.
+ *
+ * @param[in] context Context passed to the program.
+ * @param[in] program_counter The program counter (the index of the instruction to execute)
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:390]:
```suggestion
 * @param[in] program_counter The program counter (the index of the instruction to execute).
```

File: libfuzzer/libfuzz_harness.cc [hunk 8]
```diff
@@ -95,6 +341,146 @@ ubpf_vm_ptr create_ubpf_vm(const std::vector<uint8_t>& program_code)
...
+ * @param[in] stack_start The start of the stack.
+ * @param[in] stack_length The length of the stack.
+ * @param[in] register_mask The set of registers that have been modified since the start of the program.
+ * @param[in] stack_mask The set of stack locations that have been modified since the start of the program.
+ */
+void
+ubpf_debug_function(
+    void* context, int program_counter, const uint64_t registers[16], const uint8_t* stack_start, size_t stack_length, uint64_t register_mask, const uint8_t* stack_mask)
+{
+#if 1
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:401]: clean this up

Comment (Alan-Jowett) [reply libfuzzer/libfuzz_harness.cc]: Switched this to a CMake option for now. Once https://github.com/vbpf/ebpf-verifier/pull/729 is merged we can decide if we want this to be always on.

File: libfuzzer/libfuzz_harness.cc [hunk 9]
```diff
@@ -95,6 +381,185 @@ ubpf_vm_ptr create_ubpf_vm(const std::vector<uint8_t>& program_code)
     return vm;
 }
 
+/**
+ * @brief Classify the given register value as packet, context, stack, map, or unknown.
```
Comment (hawkinsw) [inline libfuzzer/libfuzz_harness.cc:385]:
```suggestion
 * @brief Classify the given address as packet, context, stack, map, or unknown.
```

File: vm/inc/ubpf.h [hunk 10]
```diff
@@ -571,13 +579,18 @@ extern "C"
      * @param[in] registers Array of 11 registers representing the VM state.
      * @param[in] stack_start Pointer to the beginning of the stack.
      * @param[in] stack_length Size of the stack in bytes.
+     * @param[in] register_mask Bitmask of registers that have been modified since the start of the program.
+     * @param[in] stack_mask_start Bitmask of the stack that has been modified since the start of the program.
+     *  Each set bit represents 1 byte of the stack that has been modified.
```
Comment (hawkinsw) [inline vm/inc/ubpf.h:584]: Is it important for the user to know whether the top of the stack is at the lower address in this bitmask or the higher address?

**Review by hawkinsw — COMMENTED** [review]
I will review more as soon as I can, but I thought that I could send along these tiny little bits of feedback now! Thank you for doing this work -- I think that it is going to be tremendously helpful!

File: libfuzzer/libfuzz_harness.cc [hunk 11]
```diff
@@ -105,19 +570,28 @@ ubpf_vm_ptr create_ubpf_vm(const std::vector<uint8_t>& program_code)
  * @return true if the program executed successfully.
  * @return false if the program failed to execute.
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:571]:
```suggestion
 * @retval true if the program executed successfully.
 * @retval false if the program failed to execute.
```

File: libfuzzer/libfuzz_harness.cc [hunk 12]
```diff
@@ -134,10 +608,17 @@ bool call_ubpf_interpreter(const std::vector<uint8_t>& program_code, std::vector
...
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:609]:
```suggestion
 * @retval true if the program executed successfully.
 * @retval false if the program failed to execute.
```

File: libfuzzer/libfuzz_harness.cc [hunk 13]
```diff
@@ -170,7 +648,8 @@ bool call_ubpf_jit(const std::vector<uint8_t>& program_code, std::vector<uint8_t
  * @return true if the input buffer was successfully split.
  * @return false if the input buffer is malformed.
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:649]:
```suggestion
 * @retval true if the input buffer was successfully split.
 * @retval false if the input buffer is malformed.
```

File: libfuzzer/libfuzz_harness.cc [hunk 14]
```diff
@@ -224,7 +703,8 @@ bool split_input(const uint8_t* data, std::size_t size, std::vector<uint8_t>& pr
  * @return -1 if the input is invalid
  * @return 0 if the input is valid and processed.
```
Comment (dthaler) [inline libfuzzer/libfuzz_harness.cc:704]:
```suggestion
 * @retval -1 The input is invalid
 * @retval 0 The input is valid and processed.
```

File: vm/ubpf_vm.c [hunk 15]
```diff
@@ -521,7 +521,16 @@ ubpf_validate_shadow_register(const struct ubpf_vm* vm, uint16_t* shadow_registe
             dst_register_required = true;
             break;
         case 0xb0: // EBPF_OP_MOV
-            // Destination register is initialized.
+            // Special case:
+            // If the source register is not initialized, the destination register is not initialized.
+            // If the source register is initialized, the destination register is initialized.
+            src_register_required = false;
+            // If source register is not initialized, the destination register is not initialized.
```
Comment (dthaler) [inline vm/ubpf_vm.c:528]:
```suggestion
```
redundant with line 525

File: .gitignore [hunk 16]
```diff
@@ -115,4 +115,6 @@ dmypy.json
 .pyre/
 
 .vscode/
-.vs/
\ No newline at end of file
+.vs/
+
+corpus/
```
Comment (dthaler) [inline .gitignore:120]: Why this change?  It's not mentioned in the PR description.

Comment (Alan-Jowett) [reply .gitignore]: Removed.

**Review by hawkinsw — COMMENTED** [review]
I think that there is a problem with our CMake configuration when the developer wants to have both "regular" tests and "fuzzing" tests on Linux. In that configuration, I get a linker error because `vm/test.c`  has a `main` and then the _libfuzzer_ (which is linked) has a `main`, too. I will try to do some debugging and see whether my hunch is true.

**Comment by hawkinsw** [comment]
> I think that there is a problem with our CMake configuration when the developer wants to have both "regular" tests and "fuzzing" tests on Linux. In that configuration, I get a linker error because `vm/test.c` has a `main` and then the _libfuzzer_ (which is linked) has a `main`, too. I will try to do some debugging and see whether my hunch is true.

I have confirmed that this "cross up" is the problem -- when the user (on Linux) has conformance tests and fuzzing tests enabled at the same time, there is a linker error.

There also seems to be a need to do an additional sanity test in `cmake/settings.cmake` -- it has a test for whether the `CC` is set to `clang` (for fuzzing) but does not guarantee that `CXX` is also properly set. I accidentally set `CC` to `clang` but had `CXX` set to `g++` and I got a linker error re: 

```
c++: error: unrecognized argument to ‘-fsanitize=’ option: ‘fuzzer’
```

It was easily correctable, but I loved what you did with the detection for the C compiler and thought we could do the same with the C++ compiler.
